data/impl/RecordValidator.js

```diff
@@ -46,71 +50,89 @@ export class RecordValidator extends HoistBase {
     /** @return {RecordErrorMap} - map of field names -> field-level errors. */
     @computed.struct
     get errors() {
-        return this.getErrorMap();
+        return this._fieldErrors ?? {};
     }
 
     /** @return {number} - count of all validation errors for the record. */
     @computed
     get errorCount() {
-        return sumBy(this._validators, 'errorCount');
+        return flatten(values(this._fieldErrors)).length;
     }
 
     /** @return {boolean} - true if any fields are currently recomputing their validation state. */
     @computed
     get isPending() {
-        return some(this._validators, it => it.isPending);
+        return this._validationTask.isPending;
     }
 
-    _validators = [];
-
     /**
      * @param {Object} c - RecordValidator configuration.
      * @param {StoreRecord} c.record - record to validate
      */
     constructor({record}) {
-        super();
-        makeObservable(this);
         this.record = record;
-
-        const {fields} = this.record.store;
-        this._validators = fields.map(field => new RecordFieldValidator({record, field}));
+        makeObservable(this);
     }
 
     /**
      * Recompute validations for the record and return true if valid.
      * @returns {Promise<boolean>}
      */
     async validateAsync() {
-        const promises = map(this._validators, v => v.validateAsync());
-        await Promise.all(promises);
+        let runId = ++this._validationRunId,
+            fieldErrors = {},
+            {record} = this,
+            fieldsToValidate = record.store.fields.filter(it => !isEmpty(it.rules));
+
+        const promises = fieldsToValidate.flatMap(field => {
+            fieldErrors[field.name] = [];
+            return field.rules.map(async (rule) => {
+                const result = await this.evaluateRuleAsync(record, field, rule);
+                fieldErrors[field.name].push(result);
+            });
+        });
+        await Promise.all(promises).linkTo(this._validationTask);
+
+        if (runId !== this._validationRunId) return;
+        fieldErrors = mapValues(fieldErrors, it => compact(flatten(it)));
+
+        runInAction(() => this._fieldErrors = fieldErrors);
+
         return this.isValid;
     }
 
     /** @return {ValidationState} - the current validation state for the record. */
     getValidationState() {
         const VS = ValidationState,
-            states = map(this._validators, v => v.validationState);
-        if (states.includes(VS.NotValid)) return VS.NotValid;
-        if (states.includes(VS.Unknown)) return VS.Unknown;
-        return VS.Valid;
-    }
+            {_fieldErrors} = this;
 
-    /** @return {RecordErrorMap} - map of field names -> field-level errors. */
-    getErrorMap() {
-        const ret = {};
-        this._validators.forEach(v => ret[v.id] = v.errors);
-        return ret;
+        if (_fieldErrors === null) return VS.Unknown; // Before executing any rules
```

Comment:
Looks like we initialize this to `{}` - when will it be `null`?

Reply:
oh!  good catch.